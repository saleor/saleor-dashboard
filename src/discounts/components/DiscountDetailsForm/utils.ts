import { Rule } from "@dashboard/discounts/models";
import { PromotionRuleDetailsFragment } from "@dashboard/graphql";

export const filterRules = (
  promotionRules: PromotionRuleDetailsFragment[],
  formRules: Rule[],
  dirtyRulesIndexes: string[],
): Rule[] => {
  if (!promotionRules) {
    return [];
  }

  return formRules.filter((rule, index) => {
    // Selected only dirty rules to update
    if (promotionRules.find(r => r.id === rule.id)) {
      return dirtyRulesIndexes.includes(index.toString());
    }

    // Keep all new added rules
    return true;
  });
};
