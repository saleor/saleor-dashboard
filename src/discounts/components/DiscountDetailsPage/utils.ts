import { Rule } from "@dashboard/discounts/types";
import { PromotionDetailsFragment } from "@dashboard/graphql";

export const filterRules = (
  promotionRules: PromotionDetailsFragment["rules"],
  formRules: Rule[],
  dirtyRulesIndexes: string[],
): Rule[] => {
  if (!promotionRules) {
    return [];
  }

  return formRules.filter((rule, index) => {
    // Selected only dirty rules to update
    if (promotionRules.find(r => r.id === rule.id)) {
      return dirtyRulesIndexes.includes(index.toString());
    }

    // Keep all new added rules
    return true;
  });
};
