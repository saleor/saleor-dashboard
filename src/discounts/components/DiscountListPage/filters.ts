import { IFilter } from "@dashboard/components/Filter";
import { MultiAutocompleteChoiceType } from "@dashboard/components/MultiAutocompleteSelectField";
import { DiscountStatusEnum, DiscountValueTypeEnum } from "@dashboard/graphql";
import { FilterOpts, MinMax } from "@dashboard/types";
import {
  createDateField,
  createOptionsField,
} from "@dashboard/utils/filters/fields";
import { defineMessages, IntlShape } from "react-intl";

export enum SaleFilterKeys {
  saleType = "saleType",
  started = "started",
  status = "status",
  channel = "channel",
}

export interface SaleListFilterOpts {
  saleType: FilterOpts<DiscountValueTypeEnum>;
  started: FilterOpts<MinMax>;
  status: FilterOpts<DiscountStatusEnum[]>;
  channel: FilterOpts<string> & { choices: MultiAutocompleteChoiceType[] };
}

const messages = defineMessages({
  active: {
    id: "AnqH4p",
    defaultMessage: "Active",
    description: "sale status",
  },
  channel: {
    id: "1BNKCZ",
    defaultMessage: "Channel",
    description: "sale channel",
  },
  expired: {
    id: "RBxYJf",
    defaultMessage: "Expired",
    description: "sale status",
  },
  fixed: {
    id: "XDBeA+",
    defaultMessage: "Fixed amount",
    description: "discount type",
  },
  percentage: {
    id: "s17U7u",
    defaultMessage: "Percentage",
    description: "discount type",
  },
  scheduled: {
    id: "BanAhF",
    defaultMessage: "Scheduled",
    description: "sale status",
  },
  started: {
    id: "zjHH6b",
    defaultMessage: "Started",
    description: "sale start date",
  },
  status: {
    id: "SpngiS",
    defaultMessage: "Status",
    description: "sale status",
  },
  type: {
    id: "KHZlmi",
    defaultMessage: "Discount Type",
  },
});

export function createFilterStructure(
  intl: IntlShape,
  opts: SaleListFilterOpts,
): IFilter<SaleFilterKeys> {
  return [
    {
      ...createOptionsField(
        SaleFilterKeys.channel,
        intl.formatMessage(messages.channel),
        [opts.channel.value],
        false,
        opts.channel.choices,
      ),
      active: opts.channel.active,
    },
    {
      ...createDateField(
        SaleFilterKeys.started,
        intl.formatMessage(messages.started),
        opts.started.value,
      ),
      active: opts.started.active,
    },
    {
      ...createOptionsField(
        SaleFilterKeys.status,
        intl.formatMessage(messages.status),
        opts.status.value,
        true,
        [
          {
            label: intl.formatMessage(messages.active),
            value: DiscountStatusEnum.ACTIVE,
          },
          {
            label: intl.formatMessage(messages.expired),
            value: DiscountStatusEnum.EXPIRED,
          },
          {
            label: intl.formatMessage(messages.scheduled),
            value: DiscountStatusEnum.SCHEDULED,
          },
        ],
      ),
      active: opts.status.active,
    },
    {
      ...createOptionsField(
        SaleFilterKeys.saleType,
        intl.formatMessage(messages.type),
        [opts.saleType.value],
        false,
        [
          {
            label: intl.formatMessage(messages.fixed),
            value: DiscountValueTypeEnum.FIXED,
          },
          {
            label: intl.formatMessage(messages.percentage),
            value: DiscountValueTypeEnum.PERCENTAGE,
          },
        ],
      ),
      active: opts.saleType.active,
    },
  ];
}
