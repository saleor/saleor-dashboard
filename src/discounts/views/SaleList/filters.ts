import { IFilterElement } from "@saleor/components/Filter";
import {
  SaleFilterKeys,
  SaleListFilterOpts
} from "@saleor/discounts/components/SaleListPage";
import { findValueInEnum, joinDateTime, maybe } from "@saleor/misc";
import {
  DiscountStatusEnum,
  DiscountValueTypeEnum,
  SaleFilterInput
} from "@saleor/types/globalTypes";

import {
  createFilterTabUtils,
  createFilterUtils,
  dedupeFilter,
  getGteLteVariables,
  getMinMaxQueryParam,
  getMultipleEnumValueQueryParam,
  getSingleEnumValueQueryParam
} from "../../../utils/filters";
import {
  SaleListUrlFilters,
  SaleListUrlFiltersEnum,
  SaleListUrlFiltersWithMultipleValues,
  SaleListUrlQueryParams
} from "../../urls";

export const SALE_FILTERS_KEY = "saleFilters";

export function getFilterOpts(params: SaleListUrlFilters): SaleListFilterOpts {
  return {
    saleType: {
      active: !!maybe(() => params.type),
      value: maybe(() => findValueInEnum(params.type, DiscountValueTypeEnum))
    },
    started: {
      active: maybe(
        () =>
          [params.startedFrom, params.startedTo].some(
            field => field !== undefined
          ),
        false
      ),
      value: {
        max: maybe(() => params.startedTo, ""),
        min: maybe(() => params.startedFrom, "")
      }
    },
    status: {
      active: !!maybe(() => params.status),
      value: maybe(
        () =>
          dedupeFilter(
            params.status.map(status =>
              findValueInEnum(status, DiscountStatusEnum)
            )
          ),
        []
      )
    }
  };
}

export function getFilterVariables(
  params: SaleListUrlFilters
): SaleFilterInput {
  return {
    saleType:
      params.type && findValueInEnum(params.type, DiscountValueTypeEnum),
    search: params.query,
    started: getGteLteVariables({
      gte: joinDateTime(params.startedFrom),
      lte: joinDateTime(params.startedTo)
    }),
    status:
      params.status &&
      params.status.map(status => findValueInEnum(status, DiscountStatusEnum))
  };
}

export function getFilterQueryParam(
  filter: IFilterElement<SaleFilterKeys>
): SaleListUrlFilters {
  const { name } = filter;

  switch (name) {
    case SaleFilterKeys.saleType:
      return getSingleEnumValueQueryParam(
        filter,
        SaleListUrlFiltersEnum.type,
        DiscountValueTypeEnum
      );

    case SaleFilterKeys.started:
      return getMinMaxQueryParam(
        filter,
        SaleListUrlFiltersEnum.startedFrom,
        SaleListUrlFiltersEnum.startedTo
      );

    case SaleFilterKeys.status:
      return getMultipleEnumValueQueryParam(
        filter,
        SaleListUrlFiltersWithMultipleValues.status,
        DiscountStatusEnum
      );
  }
}

export const {
  deleteFilterTab,
  getFilterTabs,
  saveFilterTab
} = createFilterTabUtils<SaleListUrlFilters>(SALE_FILTERS_KEY);

export const { areFiltersApplied, getActiveFilters } = createFilterUtils<
  SaleListUrlQueryParams,
  SaleListUrlFilters
>({
  ...SaleListUrlFiltersEnum,
  ...SaleListUrlFiltersWithMultipleValues
});
