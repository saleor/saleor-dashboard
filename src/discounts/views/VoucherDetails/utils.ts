import { VoucherCode } from "@dashboard/discounts/components/VoucherCodesDatagrid/types";

export const getVoucherCodesToDisplay = ({
  clientVoucherCodes,
  serverVoucherCodes,
  isServerPagination,
  freeSlotsInServerPagianationPage,
  hasServerPaginationPrevPage,
  hasClientPaginationNextPage,
  freeSlotsInClientPagianationPage,
}: {
  isServerPagination: boolean;
  clientVoucherCodes: VoucherCode[];
  serverVoucherCodes: VoucherCode[];
  freeSlotsInServerPagianationPage: number;
  hasServerPaginationPrevPage: boolean;
  hasClientPaginationNextPage: boolean;
  freeSlotsInClientPagianationPage: number;
}) => {
  if (isServerPagination) {
    // Fill missing slots using client voucher codes when there are no more server voucher codes
    // Happend when user is on navigating back and reach begining of server pagination
    return [
      ...(freeSlotsInServerPagianationPage > 0 && !hasServerPaginationPrevPage
        ? clientVoucherCodes
        : []),
      ...serverVoucherCodes,
    ];
  }

  // Fill missing slots in clinet navigation with server voucher codes
  return [
    ...clientVoucherCodes,
    ...(!hasClientPaginationNextPage && freeSlotsInClientPagianationPage > 0
      ? [...serverVoucherCodes]
      : []),
  ];
};
