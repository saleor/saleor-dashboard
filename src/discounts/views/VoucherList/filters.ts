import { IFilterElement } from "@saleor/components/Filter";
import {
  VoucherFilterKeys,
  VoucherListFilterOpts
} from "@saleor/discounts/components/VoucherListPage";
import { findValueInEnum, joinDateTime, maybe } from "@saleor/misc";
import {
  DiscountStatusEnum,
  VoucherDiscountType,
  VoucherFilterInput
} from "@saleor/types/globalTypes";

import {
  createFilterTabUtils,
  createFilterUtils,
  dedupeFilter,
  getGteLteVariables,
  getMinMaxQueryParam,
  getMultipleEnumValueQueryParam
} from "../../../utils/filters";
import {
  VoucherListUrlFilters,
  VoucherListUrlFiltersEnum,
  VoucherListUrlFiltersWithMultipleValues,
  VoucherListUrlQueryParams
} from "../../urls";

export const VOUCHER_FILTERS_KEY = "voucherFilters";

export function getFilterOpts(
  params: VoucherListUrlFilters
): VoucherListFilterOpts {
  return {
    saleType: {
      active: !!maybe(() => params.type),
      value: maybe(
        () =>
          dedupeFilter(
            params.type.map(type => findValueInEnum(type, VoucherDiscountType))
          ),
        []
      )
    },
    started: {
      active: maybe(
        () =>
          [params.startedFrom, params.startedTo].some(
            field => field !== undefined
          ),
        false
      ),
      value: {
        max: maybe(() => params.startedTo, ""),
        min: maybe(() => params.startedFrom, "")
      }
    },
    status: {
      active: !!maybe(() => params.status),
      value: maybe(
        () =>
          dedupeFilter(
            params.status.map(status =>
              findValueInEnum(status, DiscountStatusEnum)
            )
          ),
        []
      )
    },
    timesUsed: {
      active: maybe(
        () =>
          [params.timesUsedFrom, params.timesUsedTo].some(
            field => field !== undefined
          ),
        false
      ),
      value: {
        max: maybe(() => params.timesUsedTo, ""),
        min: maybe(() => params.timesUsedFrom, "")
      }
    }
  };
}

export function getFilterVariables(
  params: VoucherListUrlFilters
): VoucherFilterInput {
  return {
    discountType:
      params.type &&
      params.type.map(type => findValueInEnum(type, VoucherDiscountType)),
    search: params.query,
    started: getGteLteVariables({
      gte: joinDateTime(params.startedFrom),
      lte: joinDateTime(params.startedTo)
    }),
    status:
      params.status &&
      params.status.map(status => findValueInEnum(status, DiscountStatusEnum)),
    timesUsed: getGteLteVariables({
      gte: parseInt(params.timesUsedFrom, 0),
      lte: parseInt(params.timesUsedTo, 0)
    })
  };
}

export function getFilterQueryParam(
  filter: IFilterElement<VoucherFilterKeys>
): VoucherListUrlFilters {
  const { name } = filter;

  switch (name) {
    case VoucherFilterKeys.saleType:
      return getMultipleEnumValueQueryParam(
        filter,
        VoucherListUrlFiltersWithMultipleValues.type,
        VoucherDiscountType
      );

    case VoucherFilterKeys.started:
      return getMinMaxQueryParam(
        filter,
        VoucherListUrlFiltersEnum.startedFrom,
        VoucherListUrlFiltersEnum.startedTo
      );

    case VoucherFilterKeys.timesUsed:
      return getMinMaxQueryParam(
        filter,
        VoucherListUrlFiltersEnum.timesUsedFrom,
        VoucherListUrlFiltersEnum.timesUsedTo
      );

    case VoucherFilterKeys.status:
      return getMultipleEnumValueQueryParam(
        filter,
        VoucherListUrlFiltersWithMultipleValues.status,
        DiscountStatusEnum
      );
  }
}

export const {
  deleteFilterTab,
  getFilterTabs,
  saveFilterTab
} = createFilterTabUtils<VoucherListUrlFilters>(VOUCHER_FILTERS_KEY);

export const { areFiltersApplied, getActiveFilters } = createFilterUtils<
  VoucherListUrlQueryParams,
  VoucherListUrlFilters
>({
  ...VoucherListUrlFiltersEnum,
  ...VoucherListUrlFiltersWithMultipleValues
});
