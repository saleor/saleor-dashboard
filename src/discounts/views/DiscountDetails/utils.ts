import {
  ConditionsDetailsQuery,
  PromotionDetailsQuery,
} from "@dashboard/graphql";
import { mapEdgesToItems } from "@dashboard/utils/maps";

export const getAllConditionsIds = (
  data: PromotionDetailsQuery,
): {
  productsIds: string[];
  categoriesIds: string[];
  collectionsIds: string[];
  variantsIds: string[];
} => {
  return (data?.promotion?.rules ?? []).reduce(
    (acc, rule) => {
      rule.cataloguePredicate.OR.forEach(predicate => {
        if (predicate.productPredicate) {
          acc.productsIds.push(...predicate.productPredicate.ids);
        }
        if (predicate.categoryPredicate) {
          acc.categoriesIds.push(...predicate.categoryPredicate.ids);
        }
        if (predicate.collectionPredicate) {
          acc.collectionsIds.push(...predicate.collectionPredicate.ids);
        }
        if (predicate.variantPredicate) {
          acc.variantsIds.push(...predicate.variantPredicate.ids);
        }
      });
      return acc;
    },
    {
      productsIds: [],
      categoriesIds: [],
      collectionsIds: [],
      variantsIds: [],
    },
  );
};

export const getConditonLabels = (
  data: ConditionsDetailsQuery | undefined,
): Record<string, string> => {
  if (!data) {
    return {};
  }

  return Object.values(data).reduce((acc, value) => {
    const items = mapEdgesToItems(
      value as ConditionsDetailsQuery["categories"],
    );
    items.forEach(item => {
      acc[item.id] = item.name;
    });

    return acc;
  }, {});
};
