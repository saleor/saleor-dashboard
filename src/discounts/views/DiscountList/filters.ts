import { createFilterTabUtils } from "@dashboard/utils/filters";

export const DISCOUNT_FILTERS_KEY = "saleFilters";

export const storageUtils = createFilterTabUtils<string>(DISCOUNT_FILTERS_KEY);
