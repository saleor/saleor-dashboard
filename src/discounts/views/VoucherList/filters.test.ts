import { createFilterStructure } from "@saleor/discounts/components/VoucherListPage";
import { VoucherListUrlFilters } from "@saleor/discounts/urls";
import { date } from "@saleor/fixtures";
import {
  DiscountStatusEnum,
  VoucherDiscountType
} from "@saleor/types/globalTypes";
import { getFilterQueryParams } from "@saleor/utils/filters";
import { getExistingKeys, setFilterOptsStatus } from "@test/filters";
import { config } from "@test/intl";
import { stringify as stringifyQs } from "qs";
import { createIntl } from "react-intl";

import { getFilterQueryParam, getFilterVariables } from "./filters";

describe("Filtering query params", () => {
  it("should be empty object if no params given", () => {
    const params: VoucherListUrlFilters = {};
    const filterVariables = getFilterVariables(params);

    expect(getExistingKeys(filterVariables)).toHaveLength(0);
  });

  it("should not be empty object if params given", () => {
    const params: VoucherListUrlFilters = {
      startedFrom: date.from,
      startedTo: date.to,
      status: [DiscountStatusEnum.ACTIVE, DiscountStatusEnum.EXPIRED],
      timesUsedFrom: date.from,
      timesUsedTo: date.to,
      type: [VoucherDiscountType.FIXED, VoucherDiscountType.SHIPPING]
    };
    const filterVariables = getFilterVariables(params);

    expect(getExistingKeys(filterVariables)).toHaveLength(4);
  });
});

describe("Filtering URL params", () => {
  const intl = createIntl(config);

  const filters = createFilterStructure(intl, {
    saleType: {
      active: false,
      value: [VoucherDiscountType.FIXED, VoucherDiscountType.SHIPPING]
    },
    started: {
      active: false,
      value: {
        max: date.to,
        min: date.from
      }
    },
    status: {
      active: false,
      value: [DiscountStatusEnum.ACTIVE, DiscountStatusEnum.EXPIRED]
    },
    timesUsed: {
      active: false,
      value: {
        max: "6",
        min: "1"
      }
    }
  });

  it("should be empty if no active filters", () => {
    const filterQueryParams = getFilterQueryParams(
      filters,
      getFilterQueryParam
    );

    expect(getExistingKeys(filterQueryParams)).toHaveLength(0);
  });

  it("should not be empty if active filters are present", () => {
    const filterQueryParams = getFilterQueryParams(
      setFilterOptsStatus(filters, true),
      getFilterQueryParam
    );

    expect(filterQueryParams).toMatchSnapshot();
    expect(stringifyQs(filterQueryParams)).toMatchSnapshot();
  });
});
