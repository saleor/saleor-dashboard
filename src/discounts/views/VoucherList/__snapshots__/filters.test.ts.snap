// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filtering URL params should not be empty if active filters are present 1`] = `
Object {
  "channel": "default-channel",
  "startedFrom": "2019-12-09",
  "startedTo": "2019-12-38",
  "status": Array [
    "ACTIVE",
    "EXPIRED",
  ],
  "timesUsedFrom": "1",
  "timesUsedTo": "6",
  "type": Array [
    "FIXED",
    "SHIPPING",
  ],
}
`;

exports[`Filtering URL params should not be empty if active filters are present 2`] = `"channel=default-channel&startedFrom=2019-12-09&startedTo=2019-12-38&timesUsedFrom=1&timesUsedTo=6&status%5B0%5D=ACTIVE&status%5B1%5D=EXPIRED&type%5B0%5D=FIXED&type%5B1%5D=SHIPPING"`;
